File: src/main/scala/com/tpg/pnode/ListenerActor.scala
package com.tpg.pnode

import akka.actor.{ActorRef, Actor}
import com.github.sstone.amqp.Amqp.{Ack, Delivery}
import com.tpg.pnode.rabbit.RabbitQueueSource.RabbitMsg

/**
 * Actor binding rabbit to a stream publisher actor
 * @param producer an actor that we will forward the messages to - the actor that will act as the 'Source' of the stream
 */
class ListenerActor(producer: ActorRef) extends Actor {

  def receive = {
    case Delivery(consumerTag, envelope, properties, body) => {
      val msg = new String(body)

      // tell the stream publisher that there's a new message available
      producer ! RabbitMsg(msg)

      // tell rabbit that we took care of the message
      sender ! Ack(envelope.getDeliveryTag)
    }
    case o => println("Not handling %s message in rabbit listener actor".format(o))
  }

}

File: src/main/java/com/tpg/java/ListenerActor.java
package com.tpg.java;

import akka.actor.AbstractActor;
import akka.actor.ActorRef;
import akka.japi.pf.ReceiveBuilder;
import com.github.sstone.amqp.Amqp;
import com.tpg.java.rabbithandler.RabbitQueueSourceActor;


/**
 * Actor binding rabbit to a stream publisher actor
 */
class ListenerActor extends AbstractActor {

    /**
     *
     * @param producer an actor that we will forward the messages to - the actor that will act as the 'Source' of the stream
     */
    public ListenerActor(final ActorRef producer) {
        receive(ReceiveBuilder.
                        match(Amqp.Delivery.class, d -> {
                            String msg = new String(d.body());

                            // tell the stseam publisher that there's a new message available
                            producer.tell(new RabbitQueueSourceActor.RabbitMsg(msg), self());

                            // tell rabbit that we took care of the message
                            sender().tell(new Amqp.Ack(d.envelope().getDeliveryTag()), self());
                        }).
                        matchAny(o -> {
                            System.out.println(String.format("Not handling %s message in rabbit listener actor", o));
                        }).build()
        );

    }

}